### PR TITLE
Compile transupp.c as part of the library

### PR DESCRIPTION
**Pull request by L-as**
Perhaps this is done for a good reason, but I don't see how else you'd easily use this functionality.

### CODE REVIEW

**Comment by Ericson2314** [comment]
See https://github.com/NixOS/nixpkgs/pull/92708 for freeimage which needs this (if it's not to vendor things). Also note other distros do this with freeimage too.

**Comment by dcommander** [comment]
**transupp.c** was introduced with libjpeg v6b nearly 23 years ago as an application-private API for the jpegtran application, which (along with cjpeg and djpeg) serves as an example of libjpeg API usage (more specifically, in the case of jpegtran, the `jpeg_read_coefficients()` and `jpeg_write_coefficients()` interfaces.)  libjpeg-turbo is an ISO standard reference implementation and has a direct or indirect user base of millions.  People monitor our public APIs and ABIs for disruptive changes, so a lot more consideration would have to be taken before expanding the public libjpeg API to accommodate lossless transformation.  It isn't just a matter of compiling **transupp.c** into the libjpeg API library and then installing **transupp.h**.  The API in **transupp.c** was never intended to be public and has thus not been designed with backward compatibility and future-proofness in mind.  This five-minute patch would cause me years of headaches if I accepted it.  That's why I strongly recommend that potential contributors submit an issue first before submitting a PR.

Applications and higher-level APIs that need the functionality in **transupp.c** should just include **transupp.c**.  That's what we do within libjpeg-turbo (jpegtran and the TurboJPEG API library both include their own copies of **transupp.c** rather than expecting the libjpeg API library to provide it.)  Downstream applications and higher-level APIs could also use the lossless transform function in the TurboJPEG API, if the TurboJPEG API is otherwise suitable (it doesn't provide all of the advanced features that the libjpeg API provides.)  We can certainly talk about including lossless transform functionality in a proposed new API (refer to #313), but given this project's dire lack of funding right now, don't expect any movement on that front anytime soon.

**Comment by L-as** [comment]
Thank you for the explanation.

**Comment by dcommander** [comment]
Another note regarding the stability of **transupp.c**:

Historically, apart from cosmetic formatting changes and such, that code has not changed much relative to the implementation in libjpeg.  Most of the changes have been to incorporate lossless transform enhancements introduced in libjpeg v7+, and in fact, the application-private nature of **transupp.c** makes it a lot easier to incorporate those enhancements.  That is yet another justification for keeping **transupp.c** private, and the main justification for making it public (the ability for downstream users of the code to automatically receive bug fixes and performance enhancements) isn't applicable in this case.

In this case, it's more flexible for downstream applications and frameworks to incorporate whichever version of **transupp.c** they want (even the original version from libjpeg v6b should work, if the application or framework doesn't need the expanded lossless transform functionality from libjpeg v7+.)  That gives downstream project maintainers the ability to freeze their particular implementations of lossless JPEG transformation with a particular feature set while retaining the ability to receive JPEG codec bug fixes and performance enhancements through the libjpeg API library.  Also, including **transupp.c** in the libjpeg API library would create symbol conflicts for downstream projects that already integrated **transupp.c**.  That would force downstream projects to alter their build systems based on the version of libjpeg-turbo they are using, which would create a lot more downstream pain than this PR purports to eliminate.

**Comment by L-as** [comment]
@dcommander how is one supposed to use transupp.c downstream if it depends on jcopy_block_row, which seems to be internal?

**Comment by dcommander** [comment]
Again, you would use it the same way we use it:

https://github.com/libjpeg-turbo/libjpeg-turbo/blob/master/sharedlib/CMakeLists.txt#L84-L86

If you try that and it doesn't work, then let me know, but I only fix real issues, not hypothetical ones.

**Comment by L-as** [comment]
@dcommander I can do that and it works, but only if I have access to the source code when building: transupp.c depends on some functionality that needs access to unexported headers.

**Comment by dcommander** [comment]
Right, so you have to copy **transupp.h**, **jpegcomp.h**, **jinclude.h**, and **jpegint.h** into your project as well.  `jcopy_block_row()` and other functions defined in **jpegint.h** are exposed as global symbols in the libjpeg API library but not defined in the public headers, i.e. they are not part of the official API.

Please note that I did not design the interfaces in question.  I inherited them from libjpeg.  Perhaps if Tom Lane had continued working on the project past v6b, he would have eventually incorporated lossless transformation into the public API somehow, but he didn't, and the subsequent IJG maintainer has not done so either.  Thus, it is what it is, and in order to maintain backward compatibility with libjpeg, I am not in a position to change the public API unless there is a compelling reason to do so.  This is such a widespread API that one wrong move on my part could break thousands of applications instantly.  Given the fact that I am not being paid a salary to work on libjpeg-turbo (so any fixes to the library would have to be completed on my own dime) and the fact that our funding is already exhausted for 2021 (unfortunately, a major corporate donation I was expecting did not come through), I have less than zero impetus to make the changes you propose. You're going to have to do what every other downstream developer has had to do since 1998, which is to deal with the limitations in the libjpeg architecture.  I have had to deal with those limitations as well.

I know that what I propose is possible, and I know that because I've done it before.  The details are left up to you.

**Comment by dcommander** [comment]
If you are philosophically opposed to integrating **transupp.c** into your project in the manner suggested, then another option is to use the TurboJPEG API.  That API addresses the most common JPEG use cases, although it lacks some of the more advanced (not to mention esoteric) features of the libjpeg API.  Modifying the TurboJPEG API, if necessary, is a much more straightforward proposition, since it is much less widespread and also much easier to extend in a backward-compatible manner (because, unlike the libjpeg API, it doesn't rely on exposed structures.)

In short, this PR will not be integrated.  Period.  I have provided you with two alternative approaches.  Please move forward with one of those.

**Comment by Ericson2314** [comment]
Thanks for saying what to vendor.

If you are sick of freeloaders, well, know that I (I can't speak for @L-as) am volunteering for my distro. I did not work on FreeImage or make it use these private interfaces, nor am I getting any sort of revenue from FreeImage working. I'm just trying to get the most packages working for others, with the minimal amount of hacks. If `transupp.c` is in no way on a track for stabilization, that's fine, it just means there's very little way to "justify" what freeimage does or migrate away from the hacks, short of dropping FreeImage.

We'll separately package the private interfaces so other applications aren't lulled into depending on them by accident. And thereby, no new requests from other downstream libraries to expose `transupp.c` should stem from our actions.

Have a nice day and good look securing funding for the future.

**Comment by L-as** [comment]
Thanks! FYI I just didn't understand if you were supposed to also include the other files, since it seemed to be contradictive to what  you said about it being example usage from my perspective. I am not saying that you should include transupp.c, and completely understand why you shouldn't.

**Comment by dcommander** [comment]
@Ericson2314 Please do not put words in my mouth.  I never used the term "freeloaders."  I made no judgement whatsoever on the actions or desires of downstream projects.  I merely provided an explanation as to our project's situation and why that makes any disruptive modifications to it unlikely unless funding is provided to develop and support those modifications.  It is what it is.  I don't understand your comment about stabilization, though.  I thought I was quite clear above when I said that **transupp.c** is stable.  The only changes we've made to it, relative to libjpeg, are cosmetic.  The interface between **transupp.c** and the libjpeg API library has not changed in 23 years, so you could in fact use the original version of **transupp.c** from libjpeg v6b if you prefer.  By including **transupp.c** downstream, you are in control of its stability.  Once you've integrated it into your project, you don't ever need to change it if it's working properly.

@L-as Yes, I agree that it's not the most intuitive thing in the world.  It is what it is.

**Comment by Ericson2314** [comment]
@dcommander Sorry I was conflating "public" and "stable".

One last idea: since `transupp.c` is de-facto stable anyways, what would you say to adding a CMake option (off by default) to conditionally make it part of the library? Then there's no risk of breakage for anyone that doesn't opt-in.

**Comment by dcommander** [comment]
No, because that would be asking me to officially support a hack, which isn't good for the long-term health of the project.  (I did mention that libjpeg-turbo is an ISO reference implementation?  https://www.iso.org/standard/75845.html.)  Inevitably, such a feature would not receive much attention, so the likelihood that it would break would be high.  And few people would use it, anyhow.  I have basically no funding for libjpeg-turbo in 2021, because an expected large corporate contribution didn't come through.  Thus, I'm going to be lucky if this project even makes it through the year.  What limited resources I have are going to have to be focused on high-priority bug fixes.  Just integrate **transupp.c** into your project, please.  No other solutions are possible right now.
